doc/bu-xthin-protocol.md

```diff
@@ -0,0 +1,382 @@
+
+    CBlockHeader header;                        // block header
+    std::vector<uint256> vTxHashes;             // vector of all transactions id's in the block
+    std::vector<CTransaction> vMissingTx;       // vector of transactions that did not match the bloom filter and the requesting peer needs
+
+
+
+
+
+### **NetMsgType::GET_XBLOCKTX**
+  
+This message is used to re-request any missing transactions that were not in the mempool but are still needed in order to reconstruct
+the block.
+
+Construction of `GET_XBLOCKTX` is as follows:
+
+    uint256 blockhash;                          // the block hash
+    std::set<uint64_t> setCheapHashesToRequest; // 64bit transactions hashes to be requested that were missing from the mempool
+
+
+
+
+### **NetMsgType::XBLOCKTX**
+ 
+This message is used to send back the transactions that were re-requested and missing from the requesting peer's memory pool
+
+Construction of `XBLOCKTX` is as follows:
+
+    uint256 blockhash;                          // the block hash
+    std::vector<CTransaction> vMissingTx;       // missing transactions to be returned
+
+
+
+
+Order and logic of P2P message requests
+---------------------------------------
+
+- `NODE A:`  receives block announcement from NODE B by an INV message and then creates and issues a GET_XTHIN to NODE B
+
+- `NODE B:`  receives GET_XTHIN and loads the attached bloom filter and creates an XTHINBLOCK which is sent back to NODE A
+         (the xthinblock is checked before sending back to see if a hash collision is within the block, if so then send a THINBLOCK instead)
+
+- `NODE A:`  receives an XTHINBLOCK with 64 bit txn hashes
+    - if there is a hash collision with the mempool then issue GETDATA(MSG_THINBLOCK) to NODE B
+        - `NODE B:` return a THINBLOCK with full 256 bit txn hashes
+        - `NODE A:` receives THINBLOCK and tries to reconstruct block
+            - if block is still incomplete then 
+                - `NODE A:` issues a GETDATA(MSG_BLOCK) for the full block
+
+    - else if missing transactions then re-request any missing transactions which can happen with a bloom filter false positive match.
+        - `NODE A:` issues a GET_XBLOCKTX to NODE B
+        - `NODE B:` returns an XBLOCKTX
+        - `NODE A:` receives XBLOCKTX and completes the block
+
+- `NODE A:` Accepts the final full block, re-constructed xthinblock or re-constructed thinblock
+
+
+
+The Importance of the Orphan Cache
+----------------------------------
+
+It was found that the orphan cache can play an important role in the "thinness" of an xthinblock.  The current core configuration for orphan caching
+is broken.  It is too small and evictions happen to frequently to be of any use.  We found two things of major importance in relation to xthinblocks
+and those are:
+
+- do not purge orphans when a node disconnects.  Currently in Bitcoin Core, when a node disconnects then all orphans that were received from it are
```

Comment:
IMO mentioning Core is irrelevant. Also, it may be useful to mention what orphans are.